### PR TITLE
Add initial implementation of the Avro Serializer for Confluent Schema Registry Connector

### PR DESCRIPTION
**Pull request by Nuvindu**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,159 @@
...
+   string schema = string `
+      {
+         "type": "int",
+         "name" : "value", 
+         "namespace": "data"
+      }`;
+
+   int value = 1;
+   byte[] bytes = check avroSerDes.serialize(schema, value, "subject");
+   int number = check avroSerDesClient.deserialize(bytes);
```
Comment (ayeshLK) [inline README.md]:
```suggestion
   int number = check avroSerDes.deserialize(bytes);
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,159 @@
...
+
+For more information, go to the [contribution guidelines](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md).
+
+## Code of conduct
+
+All contributors are encouraged to read the [Ballerina Code of Conduct](https://ballerina.io/code-of-conduct).
+
+## Useful links
+
+- Discuss code changes of the Ballerina project in [ballerina-dev@googlegroups.com](mailto:ballerina-dev@googlegroups.com).
```
Comment (ayeshLK) [inline README.md:158]: API doc link is missing here right ?

File: ballerina/Module.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   string schema = string `
+      {
+         "type": "int",
+         "name" : "value", 
+         "namespace": "data"
+      }`;
+
+   int value = 1;
+   byte[] bytes = check avroSerDes.serialize(schema, value, "subject");
+   int number = check avroSerDesClient.deserialize(bytes);
```
Comment (ayeshLK) [inline ballerina/Module.md]:
```suggestion
   int number = check avroSerDes.deserialize(bytes);

```

File: ballerina/Package.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+   string schema = string `
+      {
+         "type": "int",
+         "name" : "value", 
+         "namespace": "data"
+      }`;
+
+   int value = 1;
+   byte[] bytes = check avroSerDes.serialize(schema, value, "subject");
+   int number = check avroSerDesClient.deserialize(bytes);
```
Comment (ayeshLK) [inline ballerina/Package.md]:
```suggestion
   int number = check avroSerDes.deserialize(bytes);

```

File: ballerina/Package.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+cavroserdes:Client avroSerDes = check new ({
+    baseUrl,
+    identityMapCapacity,
+    originals,
+    headers
+});
+```
+
+### Step 3: Invoke the connector operation
```
Comment (ayeshLK) [inline ballerina/Package.md]: We usually add 4 steps right ? I think as the fourth step we add `Step 4: Run the Ballerina application`. Could you please check ?

File: ballerina/client.bal [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    #
+    # + data - Avro serialized data which includes the schema id
+    # + targetType - Default parameter use to infer the user specified type
+    # + return - A deserialized data with the given type or else an `cavroserdes:Error`
+    remote isolated function deserialize(byte[] data, typedesc<anydata> targetType = <>)
+        returns targetType|Error = @java:Method {
+        'class: "io.ballerina.lib.confluent.avro.serdes.AvroSerializer"
+    } external;
+
+    private isolated function toBytes(int id) returns byte[]|error = @java:Method {
```
Comment (ayeshLK) [inline ballerina/client.bal]: Do we need to explicitly mark this `private` ? I think anyway these functions would be package-private if we do not have any access modifier.

File: ballerina/Ballerina.toml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+[package]
+org = "ballerinax"
+name = "confluent.cavroserdes"
+version = "0.1.0"
+authors = ["Ballerina"]
+export=["confluent.cavroserdes"]
+keywords = ["confluent", "schema_registry", "avro", "serdes"]
+repository = "https://github.com/ballerina-platform/module-ballerinax-confluent.cregistry"
```
Comment (ThisaruGuruge) [inline ballerina/Ballerina.toml]: Incorrect repository URL.

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java]:
```suggestion
 * Copyright (c) 2024, WSO2 LLC. (http://www.wso2.org)
```

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import io.ballerina.runtime.api.types.UnionType;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BTypedesc;
+
+import java.nio.ByteBuffer;
+
+import static io.ballerina.lib.confluent.avro.serdes.ModuleUtils.getModule;
+
+public class AvroSerializer {
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java:39]: missing java docs

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    private static final String FUNCTION = "deserializeData";
+
+    public static final StrandMetadata EXECUTION_STRAND = new StrandMetadata(
+            getModule().getOrg(),
+            getModule().getName(),
+            getModule().getMajorVersion(),
+            FUNCTION);
+
+    public static Object deserialize(Environment env, BObject kafkaSerDes, BArray data,
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java:48]: It seems that the `AvroSerializer` class only contains a  `deserialize()` method. Is that intentional?

File: ballerina/client.bal (hunk 6 above)
Comment (ThisaruGuruge) [inline ballerina/client.bal]: Any reason for this to be a class method?

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/ExecutionCallback.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+public class ExecutionCallback implements Callback {
+    private final Future future;
+    private final BTypedesc typeDesc;
+
+    ExecutionCallback(Future future, BTypedesc typeDesc) {
+        this.future = future;
+        this.typeDesc = typeDesc;
+    }
+    @Override
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/confluent/avro/serdes/ExecutionCallback.java]:
```suggestion
    }
    
    @Override
```

File: ballerina/Package.md (hunk 5 above)
Comment (Nuvindu) [reply ballerina/Package.md]: Fixed in [0fca32b](https://github.com/ballerina-platform/module-ballerinax-confluent.cavroserdes/pull/1/commits/0fca32b45cc1e6ccca077069fa1d9b01255ed71b)

File: README.md (hunk 2 above)
Comment (Nuvindu) [reply README.md:158]: Added in [0fca32b](https://github.com/ballerina-platform/module-ballerinax-confluent.cavroserdes/pull/1/commits/0fca32b45cc1e6ccca077069fa1d9b01255ed71b)

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java (hunk 10 above)
Comment (Nuvindu) [reply native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java:48]: Yes, here the requirement of the `deserialize()` method is to return the user-inferred type and we need to use a Java API for that

File: ballerina/client.bal (hunk 6 above)
Comment (Nuvindu) [reply ballerina/client.bal]: Fixed in [c88f651](https://github.com/ballerina-platform/module-ballerinax-confluent.cavroserdes/pull/1/commits/c88f6516500d621e31ed52a61a1efb584ef68aaf)

**Review by ThisaruGuruge — APPROVED** [review]
LGTM

File: native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java (hunk 10 above)
Comment (NipunaRanasinghe) [reply native/src/main/java/io/ballerina/lib/confluent/avro/serdes/AvroSerializer.java:48]: Yeah make sense. What I meant was, if the intention is to perform only desirialization in here, the class name might not correctly reflect what it does. Wouldn't a name like `AvroDeserializer` be more appropriate in that case?

**Pull request merged**